gateway-sp-comms/src/recv_handler/request_response_map.rs

```diff
@@ -0,0 +1,170 @@
+
+        // we always clean up `self.requests` after receiving a response (or
+        // timing out), so we should never see request ID reuse even if they
+        // roll over. assert that here to ensure we don't get mysterious
+        // misbehavior if that turns out to be incorrect
+        assert!(old.is_none(), "request ID reuse");
+
+        async move {
+            // wait for someone to call `ingest_response()` with our request id
+            match (&mut wrapped_rx.rx).await {
+                Ok(inner_result) => inner_result,
+                Err(_recv_error) => {
+                    // we hold the sending half in `self.requests` until either
+                    // `wrapped_rx`'s `Drop` impl removes it (in which case
+                    // we're not running anymore), or it's consumed to send the
+                    // result to us. receiving therefore can't fail
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    /// Ingest a response, which will cause the corresponding
+    /// `wait_for_response()` future to be fulfilled (if it exists).
+    pub(super) fn ingest_response(
+        &self,
+        key: &K,
+        response: T,
+    ) -> ResponseIngestResult {
+        // get the sending half of the channel created in `wait_for_response()`,
+        // if it exists (i.e., `wait_for_response()` was actually called with
+        // `key`, and the returned future from it hasn't been dropped)
+        let tx = match self.requests.lock().unwrap().remove(key) {
+            Some(tx) => tx,
+            None => return ResponseIngestResult::UnknownRequestId,
+        };
+
+        // we got `tx`, so the receiving end existed a moment ago, but there's a
+        // race here where it could be dropped before we're able to send
+        // `result` through, so we can't unwrap this send; we treat this failure
+        // the same as not finding `tx`, because if we had tried to get `tx` a
+        // moment later it would not have been there.
+        match tx.send(response) {
+            Ok(()) => ResponseIngestResult::Ok,
+            Err(_) => ResponseIngestResult::UnknownRequestId,
+        }
+    }
+}
+
+/// `RemoveOnDrop` is a light wrapper around a [`Receiver`] that removes the
+/// associated key from the `HashMap` it's given once the response has been
+/// received (or the `RemoveOnDrop` itself is dropped).
+#[derive(Debug)]
+struct RemoveOnDrop<'a, K, T>
+where
+    K: Hash + Eq,
+{
+    rx: Receiver<T>,
+    requests: &'a Mutex<HashMap<K, Sender<T>>>,
+    key: K,
+}
+
+impl<K, T> Drop for RemoveOnDrop<'_, K, T>
+where
+    K: Hash + Eq,
+{
+    fn drop(&mut self) {
+        // we don't care to check the return value here; this will be `Some(_)`
+        // if we're being dropped before a response has been received and
+        // forwarded on to our caller, and `None` if not (because a caller will
+        // have already extracted the sending channel)
```

Comment:
Would it be valuable to leave behind some contents of `requests` in this case that we could reap later? The ostensible benefit would be that we could distinguish a timeout from a logic error. Or would this potentially add more work than it saves?

Reply:
I'm not sure what we would leave behind, and/or I'm not sure how we would distinguish between a timeout and a logic error that manifested as an undelivered message (i.e., a timeout). Did you have something specific in mind?

Reply:
Sorry for the delayed reply. I was imagining changing the type of `requests` to something like `Mutex<HashMap<K, MaybeTimedOut<Sender<T>>>>` and then in `drop` changing it to `MaybeTimedOut::TimedOut(SystemTime::now)`... that said, it sounds like a pain in the ass to then reap etc. Sorry for even saying anything... :-)